atomix/cluster/src/main/java/io/atomix/raft/partition/RaftPartitionGroup.java

```diff
@@ -250,11 +251,51 @@ public String toString() {
       for (int j = 0; j < count; j++) {
         membersForPartition.add(sorted.get((i + j) % length));
       }
-      metadata.add(new PartitionMetadata(partitionId, membersForPartition));
+      final var primary = sorted.get(i % length);
+      final var priority =
+          getPriorities(
+              partitionId, membersForPartition, primary, sorted.size(), partitions.size());
+      metadata.add(
+          new PartitionMetadata(partitionId, membersForPartition, priority, priority.get(primary)));
     }
     return metadata;
   }
 
+  private Map<MemberId, Integer> getPriorities(
+      final PartitionId partitionId,
+      final List<MemberId> membersForPartition,
+      final MemberId primary,
+      final int clusterSize,
+      final int replicationFactor) {
+    final Map<MemberId, Integer> priority = new HashMap<>();
+    final int lowestPriority = 1;
+
+    priority.put(primary, replicationFactor);
+    // To ensure that secondary priorities are distributed evenly, we alternate the nodes for which
+    // second priority is assigned. Example, clusterSize = 3 partitionCount = 12. Node 0 has highest
+    // priority (=3) for partition 1,4,7 and 9. For partition 1 and 7, node 1 gets priority 2. For
+    // partition 4 and 9, node 2 gets priority 2. This is done so that if node 0 dies, the
+    // leadership is evenly distributed on the rest of the followers.
+    if (partitionId.id() / clusterSize % 2 == 0) {
+      int nextPriority = lowestPriority;
+      for (final MemberId member : membersForPartition) {
+        if (!member.equals(primary)) {
+          priority.put(member, nextPriority);
+          nextPriority++;
+        }
+      }
+    } else {
+      int nextPriority = replicationFactor - 1;
+      for (final MemberId member : membersForPartition) {
+        if (!member.equals(primary)) {
+          priority.put(member, nextPriority);
+          nextPriority--;
+        }
+      }
+    }
+    return priority;
```

Comment:
I think we can extract that into a separate class and write an unit test for it.

Reply:
:+1: